### PR TITLE
Add ck's streamk and splitk gemm impl

### PR DESCRIPTION
**Pull request by cloudhan**

### CODE REVIEW

File: onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh [hunk 1]
```diff
@@ -69,6 +71,82 @@ auto GetCKGemmTypeStringAndOps() {
...
+
+  std::vector<std::pair<std::string, Op<GemmParams<T>>>> ret;
+  for (auto num_split : {4, 16, 64}) {
+    auto instances = InstanceFactory::GetInstances();
+    for (auto&& impl : instances) {
+      auto type_string = impl->GetTypeString() + "_SplitK" + std::to_string(num_split);
+      auto invoker = impl->MakeInvokerPointer();
+      auto ck_gemm_op = [num_split, impl = std::move(impl), invoker = std::move(invoker)](const GemmParams<T>* params) -> Status {
+        TUNABLE_OP_RETURN_UNSUPPORTED_ARGUMENT_IF(
+            params->k < 128 * num_split, "k=", num_split," is too small, it makes no sense to use this split-k gemm.");
```
Comment (PeixuanZuo) [inline onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh:126]: This log should be `"k=", 128 * num_split," is too small`

File: onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh [hunk 2]
```diff
@@ -69,6 +71,82 @@ auto GetCKGemmTypeStringAndOps() {
...
+
+  std::vector<std::pair<std::string, Op<GemmParams<T>>>> ret;
+  for (auto num_split : {4, 16, 64}) {
+    auto instances = InstanceFactory::GetInstances();
+    for (auto&& impl : instances) {
+      auto type_string = impl->GetTypeString() + "_SplitK" + std::to_string(num_split);
+      auto invoker = impl->MakeInvokerPointer();
+      auto ck_gemm_op = [num_split, impl = std::move(impl), invoker = std::move(invoker)](const GemmParams<T>* params) -> Status {
+        TUNABLE_OP_RETURN_UNSUPPORTED_ARGUMENT_IF(
+            params->k < 128 * num_split, "k=", params->k, " is too small, it makes no sense to use this split-k gemm.");
```
Comment (mindest) [inline onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh:126]: Is `k >= 128 * num_split` a hard constraint in CK? And I suggest the message be more informative like

```suggestion
            params->k < 128 * num_split, "k (", params->k, ") < 128 * num_split (", num_split, ")");
```

File: onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh [hunk 3]
```diff
@@ -69,6 +71,82 @@ auto GetCKGemmTypeStringAndOps() {
...
+auto GetCKSplitKGemmTypeStringAndOps() {
+  using CKDataType = typename CKDataTypeAdaptor<T>::type;
+  using DeviceGemm = ck::tensor_operation::device::DeviceGemmSplitK<
+      ALayout, BLayout, Row,
+      CKDataType, CKDataType, CKDataType,
+      Nop, Nop, Nop>;
+  using InstanceFactory = ck::tensor_operation::device::instance::DeviceOperationInstanceFactory<DeviceGemm>;
+
+  std::vector<std::pair<std::string, Op<GemmParams<T>>>> ret;
+  for (auto num_split : {4, 16, 64}) {
```
Comment (mindest) [inline onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh:119]: Are there only these three split sizes or there will be more? nit: num_split -> num_splits

File: onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh [hunk 4]
```diff
@@ -69,6 +71,82 @@ auto GetCKGemmTypeStringAndOps() {
...
+  std::vector<std::pair<std::string, Op<GemmParams<T>>>> ret;
+  for (auto&& impl : InstanceFactory::GetInstances()) {
+    auto type_string = impl->GetTypeString();
+    auto invoker = impl->MakeInvokerPointer();
+    auto ck_gemm_op = [impl = std::move(impl), invoker = std::move(invoker)](const GemmParams<T>* params) -> Status {
+      auto one = ToHipType<T>::FromFloat(1.0f);
+      auto zero = ToHipType<T>::FromFloat(0.0f);
+      TUNABLE_OP_RETURN_UNSUPPORTED_ARGUMENT_IF(
+          params->alpha != one || params->beta != zero,
+          impl->GetTypeString(), " only supports alpha == 1 and beta == 0", params->Signature());
```
[the remark this answers is not in the document]
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh:92]: Wont fix

File: onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh (hunk 2 above)
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh:126]: Wont fix

File: onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh (hunk 3 above)
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh:119]: They are only three arbitrarily choosed values, as the ck interface needs it as runtime variable. So we just try split with 4, 16 and 64 and hope there might be an instance that exhibit good performance.

File: onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh (hunk 4 above)
[the remark this answers is not in the document]
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh]: Prefer #17328

File: onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh [hunk 5]
```diff
@@ -69,6 +71,82 @@ auto GetCKGemmTypeStringAndOps() {
...
+      auto invoker = impl->MakeInvokerPointer();
+      auto ck_gemm_op = [num_split, impl = std::move(impl), invoker = std::move(invoker)](const GemmParams<T>* params) -> Status {
+        TUNABLE_OP_RETURN_UNSUPPORTED_ARGUMENT_IF(
+            params->k < 128 * num_split, "k=", num_split," is too small, it makes no sense to use this split-k gemm.");
+
+        auto one = ToHipType<T>::FromFloat(1.0f);
+        auto zero = ToHipType<T>::FromFloat(0.0f);
+        TUNABLE_OP_RETURN_UNSUPPORTED_ARGUMENT_IF(
+            params->alpha != one || params->beta != zero,
+            impl->GetTypeString(), " only supports alpha == 1 and beta == 0", params->Signature());
```
[the remark this answers is not in the document]
Comment (cloudhan) [reply onnxruntime/core/providers/rocm/tunable/gemm_ck.cuh]: Prefer #17328

**Review by mindest — APPROVED** [review]
LGTM, thanks!

**Pull request merged**